### PR TITLE
feat: plot detection probability on code patches

### LINKED ISSUE #135: request: visualise detection probability on patch
### Motivation

`deltakit` is able to plot detection probability (also called defect rate) as:

Image

(from https://deltakit-docs.riverlane.com/en/docs/examples/notebooks/google_qmem_experiment/noise_analysis.html)

It would be nice to also be able to plot that on a code patch when it makes sense. For example, when the rotated surface code is used, detectors can be "assigned" to plaquettes which leads to plots like

Image

(image from https://www.nature.com/articles/s41586-022-05434-1/figures/2)

### Proposed solution

Implement the above visualisation at least for the rotated surface code. We might want to be able to plot:

- for each round,
- an average of all the rounds,
- a median of all the rounds,
- variance of all the rounds except the first and last ones (which are expected to be outliers),
- [more ideas?]

### Potential alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dannyward630**
## Summary

- add `plot_detection_probability_on_patch` as a public plotting helper
- color the actual stabilizer plaquettes, including triangular boundary plaquettes
- support exact rounds plus mean, median, and interior-round variance aggregation
- allow plotting into an existing `Axes` for inset and composed figures
- validate empty, non-finite, out-of-range, and unmatched detector inputs

```python
fig, ax = plot_detection_probability_on_patch(
    RotatedPlanarCode(3, 3),
    detection_probabilities,
    aggregation=DetectionProbabilityAggregation.MEDIAN,
)
```

## Why

The plaquette geometry is the useful spatial signal here. Filling the same stabilizer polygons used by the code model makes boundary and interior behavior directly comparable without center markers obscuring the patch.

## Checks

- plotting suite: 45 passed
- full explorer suite: 14,510 passed
- Ruff check and format: passed
- mypy: passed for all 86 explorer source files
- pydoclint, typos, and deptry: passed
- `git diff --check` (passed)

Closes #135.

## AI disclosure

I used an LLM as a collaborative tool for repository triage, drafting tests, and reviewing edge cases. I inspected the existing plotting geometry, implemented and reviewed the code locally, and ran the checks listed above.

### CODE REVIEW

File: deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+from matplotlib.axes import Axes
+from matplotlib.colors import Normalize
+from matplotlib.figure import Figure
+from matplotlib.patches import Circle, Polygon
+
+if TYPE_CHECKING:
+    from deltakit_explorer.codes._planar_code import PlanarCode
+    from deltakit_explorer.codes._stabiliser import Stabiliser
+
+Aggregation = Literal["mean", "median", "variance"]
```
Comment (rolandriver) [inline deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py]: I would transform it as an enum.

File: deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        elif aggregation == "median":
+            value = float(np.median(values))
+        elif aggregation == "variance":
+            if values.size < 3:
+                msg = (
+                    "Variance requires at least three rounds so the first and last "
+                    "rounds can be excluded."
+                )
+                raise ValueError(msg)
+            value = float(np.var(values[1:-1]))
```
Comment (rolandriver) [inline deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py:102]: Could you please comment on this slice?

File: deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+
+def _aggregate_detection_probabilities(
+    detection_probabilities: Mapping[tuple[float, ...], Sequence[float]],
+    *,
+    aggregation: Aggregation = "mean",
+    round_index: int | None = None,
+) -> dict[tuple[float, float], float]:
+    """Reduce per-round detector probabilities to one value per coordinate."""
+    aggregated: dict[tuple[float, float], float] = {}
```
Comment (rolandriver) [inline deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py]: I would develop a custom datatype for this.

File: deltakit-explorer/tests/plotting/test_detection_probability.py
Comment (rolandriver) [inline deltakit-explorer/tests/plotting/test_detection_probability.py:1]: Some of the helper functions you've developed are not tested properly. Could you please add tests?

**Review by rolandriver — COMMENTED** [review]
Thanks @dannyward630. Some comments from me. Also: could you please post the result figures you can generate with this? I'll be helpful to assess visually what it does. Thank you.

**Comment by dannyward630** [comment]
Thanks for the review. I have rebased onto current `main` and pushed the requested revisions:

- replaced the string aggregation literal with the public `DetectionProbabilityAggregation` enum
- used the existing `Coord2D` domain type for aggregated spatial detector coordinates
- documented why variance uses `values[1:-1]` (state preparation and final measurement make the boundary rounds expected outliers)
- added direct tests for aggregation validation, plaquette vertex construction, coordinate-tolerance matching, missing ancillas, and round selection
- tightened probability validation to reject non-1D and out-of-range inputs

Here is a generated 5x5 preview showing mean aggregation and interior-round variance:

Detection probability plotted on a rotated planar-code patch

Verification after rebasing:
- full explorer suite: 14,510 passed
- plotting suite: 45 passed
- Ruff, mypy, pydoclint, typos, and deptry passed
- `git diff --check` passed

File: deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py (hunk 1 above)
Comment (dannyward630) [reply deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py]: Done. `DetectionProbabilityAggregation` is now a public enum with `MEAN`, `MEDIAN`, and `VARIANCE` members, and the plotting function accepts that enum directly.

File: deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py (hunk 2 above)
Comment (dannyward630) [reply deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py:102]: Added a comment at the slice. The first and last rounds correspond to state preparation and final measurement, so they are expected outliers; `values[1:-1]` keeps the variance focused on steady-state rounds, as requested in #135.

File: deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py (hunk 3 above)
Comment (dannyward630) [reply deltakit-explorer/src/deltakit_explorer/plotting/_detection_probability.py]: Done using the existing project domain type: the aggregated mapping is now `dict[Coord2D, float]` instead of raw coordinate tuples. This keeps the spatial coordinate contract consistent with the rest of Deltakit without introducing a duplicate coordinate class.

**Comment by rolandriver** [comment]
Hi @dannyward630 just to inform you that I have approved #271. Thank you for the time you spent on this really.

**Comment by dannyward630** [comment]
Thanks for letting me know, and thank you for the thoughtful review. Closing this in favor of #271 makes sense now that the competing implementation has converged and been approved. I appreciate the time you spent evaluating this branch.

**Comment by nelimee** [comment]
> Thanks for letting me know, and thank you for the thoughtful review. Closing this in favor of #271 makes sense now that the competing implementation has converged and been approved. I appreciate the time you spent evaluating this branch.

Thanks for being a thoughtful and nice contributor! Answering to reviews nicely, acknowledging the use of LLMs and submitting a very valid PR. Hope to see you again in the future :)
